Support object values for GDC, MDC and NDC contexts.

williamb1024 opened the pull request:
Add support for `object` values to GlobalDiagnosticsContext, MappedDiagnosticsContext and NestedDiagnosticsContext. 

Introduces `GetObject` methods for GlobalDiagnosticsContext and MappedDiagnosticsContext. The method operates in the same manner as `Get` but returns an `object` type. The existing `Get` methods have been replaced to render any object values using `ToString()`.

Introduces `TopObject` and `PopObject` methods to NestedDiagnosticsContext. These methods operate the same as `TopMessage` and `Pop`, but return `object` values rather than string values. The `TopMessage` and `Pop` methods have been replaced to use `ToString()` to convert `object` values before return.

The associated layout renderers have also been modified to properly format object values to strings before rendering context properties.

Resolution for #738

304NotModified commented:
Thanks.

Do you know what has been changed in Log4JXmlEventLayoutRenderer.cs (beside the line endings)?

williamb1024 commented:
Sorry about the line-endings, I think Git normalized the file and I just got to be the lucky one to do it. The changes are on line 261 and 265.

304NotModified commented:
I will do a manual compare at my pc. 

The git attributes has recently changed.  Dunno if we should 'touch' and commit everything.

williamb1024 commented:
I don't know if there is a way to avoid it without changing the .gitattributes to disable line ending normalization for the file.

304NotModified commented:
I don't know either.  But no problem.  I will check this, this weekend. I have to check the impact and then we will decide which version we will release it.

williamb1024 commented:
It appears that there are number of files that need EOL normalization at this point. I suspect normalizing the files will cause just as much grief as not normalizing them.

Here is the set of commands that I used to build the list. Once complete, all files should have their end of line normalized and not suffer from the "everything changed" problem that you noticed in my pull request.

```
rm .git/index     # Remove the index to force git to
git reset         # re-scan the working directory
git status        # Show files that will be normalized
git add -u
git commit -m "EOL normalization"
```

I imagine doing this will create merge conflicts for any outstanding pull requests, but I believe the same would be true if any of theses files are modified by different commits.

modified:   examples/NLogWindowsPhoneApplication/NLogWindowsPhoneApplication.csproj
modified:   examples/NLogWindowsPhoneApplication/Properties/WMAppManifest.xml
modified:   src/.nuget/NuGet.Config
modified:   src/.nuget/NuGet.targets
modified:   src/InstallNLogConfig/InstallNLogConfig.netfx45.csproj
modified:   src/InstallNLogConfig/app.config
modified:   src/NLog.Extended/NLog.Extended.netfx45.csproj
modified:   src/NLog.Extended/Targets/MessageQueueTarget.cs
modified:   src/NLog.monodevelop.sln
modified:   src/NLog.netfx35.sln
modified:   src/NLog.netfx40.sln
modified:   src/NLog.netfx45.sln
modified:   src/NLog/Common/InternalLogger.cs
modified:   src/NLog/Config/ConfigSectionHandler.cs
modified:   src/NLog/Config/LoggingConfiguration.cs
modified:   src/NLog/Config/XmlLoggingConfiguration.cs
modified:   src/NLog/Internal/Fakeables/AppDomainWrapper.cs
modified:   src/NLog/Internal/Fakeables/IAppDomain.cs
modified:   src/NLog/Internal/FileAppenders/CountingSingleProcessFileAppender.cs
modified:   src/NLog/Internal/FileAppenders/RetryingMultiProcessFileAppender.cs
modified:   src/NLog/Internal/FileAppenders/SingleProcessFileAppender.cs
modified:   src/NLog/Internal/NativeMethods.cs
modified:   src/NLog/Internal/NetworkSenders/INetworkSenderFactory.cs
modified:   src/NLog/Internal/NetworkSenders/NetworkSenderFactory.cs
modified:   src/NLog/Internal/NetworkSenders/SocketProxy.cs
modified:   src/NLog/Internal/NetworkSenders/TcpNetworkSender.cs
modified:   src/NLog/Internal/NetworkSenders/UdpNetworkSender.cs
modified:   src/NLog/LayoutRenderers/AssemblyVersionLayoutRenderer.cs
modified:   src/NLog/LayoutRenderers/BaseDirLayoutRenderer.cs
modified:   src/NLog/LayoutRenderers/EnvironmentLayoutRenderer.cs
modified:   src/NLog/LayoutRenderers/EventContextLayoutRenderer.cs
modified:   src/NLog/LayoutRenderers/EventPropertiesLayoutRenderer.cs
modified:   src/NLog/LayoutRenderers/Log4JXmlEventLayoutRenderer.cs
modified:   src/NLog/LayoutRenderers/Wrappers/ReplaceLayoutRendererWrapper.cs
modified:   src/NLog/Layouts/LayoutParser.cs
modified:   src/NLog/LogEventInfo.cs
modified:   src/NLog/LogFactory.cs
modified:   src/NLog/LogLevel.cs
modified:   src/NLog/NLog.doc.csproj
modified:   src/NLog/NLog.mono2.csproj
modified:   src/NLog/NLog.netfx35.csproj
modified:   src/NLog/NLog.netfx40.csproj
modified:   src/NLog/NLog.netfx45.csproj
modified:   src/NLog/NLog.sl5.csproj
modified:   src/NLog/Targets/EventLogTarget.cs
modified:   src/NLog/Targets/MailTarget.cs
modified:   src/NLog/Targets/NetworkTarget.cs
modified:   src/NLog/Targets/Target.cs
modified:   src/NLog/Targets/TraceTarget.cs
modified:   src/NLog/Targets/Wrappers/AsyncRequestQueue-T.cs
modified:   src/NLog/Targets/Wrappers/AsyncTargetWrapper.cs
modified:   src/NLog/Time/AccurateLocalTimeSource.cs
modified:   src/NLog/Time/AccurateUtcTimeSource.cs
modified:   src/NLog/Time/CachedTimeSource.cs
modified:   src/NLog/Time/FastLocalTimeSource.cs
modified:   src/NLog/Time/FastUtcTimeSource.cs
modified:   src/NLog/Time/TimeSource.cs
modified:   src/NLog/Time/TimeSourceAttribute.cs
modified:   src/NLog/packages.config
modified:   src/NuGet/NLog.Snippets/content/empty.txt
modified:   src/NuGet/NLog.Snippets/tools/install.ps1
modified:   tests/NLog.UnitTests/Config/TimeConfigurationTests.cs
modified:   tests/NLog.UnitTests/ConfigFileLocatorTests.cs
modified:   tests/NLog.UnitTests/GetLoggerTests.cs
modified:   tests/NLog.UnitTests/Internal/NetworkSenders/TcpNetworkSenderTests.cs
modified:   tests/NLog.UnitTests/Internal/NetworkSenders/UdpNetworkSenderTests.cs
modified:   tests/NLog.UnitTests/LayoutRenderers/BaseDirTests.cs
modified:   tests/NLog.UnitTests/LayoutRenderers/DateTests.cs
modified:   tests/NLog.UnitTests/LayoutRenderers/EnvironmentTests.cs
modified:   tests/NLog.UnitTests/LayoutRenderers/EventContextTests.cs
modified:   tests/NLog.UnitTests/LayoutRenderers/EventPropertiesTests.cs
modified:   tests/NLog.UnitTests/LayoutRenderers/Log4JXmlTests.cs
modified:   tests/NLog.UnitTests/LayoutRenderers/SpecialFolderTests.cs
modified:   tests/NLog.UnitTests/LayoutRenderers/Wrappers/ReplaceTests.cs
modified:   tests/NLog.UnitTests/Layouts/CsvLayoutTests.cs
modified:   tests/NLog.UnitTests/Layouts/SimpleLayoutParserTests.cs
modified:   tests/NLog.UnitTests/LogFactoryTests.cs
modified:   tests/NLog.UnitTests/LogManagerTests.cs
modified:   tests/NLog.UnitTests/LoggerTests.cs
modified:   tests/NLog.UnitTests/NLog.UnitTests.mono2.csproj
modified:   tests/NLog.UnitTests/NLog.UnitTests.netfx35.csproj
modified:   tests/NLog.UnitTests/NLog.UnitTests.netfx40.csproj
modified:   tests/NLog.UnitTests/NLog.UnitTests.netfx45.csproj
modified:   tests/NLog.UnitTests/NLog.UnitTests.sl5.csproj
modified:   tests/NLog.UnitTests/NLogTraceListenerTests.cs
modified:   tests/NLog.UnitTests/Targets/ConsoleTargetTests.cs
modified:   tests/NLog.UnitTests/Targets/FileTargetTests.cs
modified:   tests/NLog.UnitTests/Targets/LogReceiverWebServiceTargetTests.cs
modified:   tests/NLog.UnitTests/Targets/MailTargetTests.cs
modified:   tests/NLog.UnitTests/Targets/MessageQueueTargetTests.cs
modified:   tests/NLog.UnitTests/Targets/NetworkTargetTests.cs
modified:   tests/NLog.UnitTests/Targets/TargetTests.cs
modified:   tests/NLog.UnitTests/Targets/Wrappers/BufferingTargetWrapperTests.cs
modified:   tests/NLog.UnitTests/Targets/Wrappers/FallbackGroupTargetTests.cs
modified:   tests/NLog.UnitTests/Targets/Wrappers/PostFilteringTargetWrapperTests.cs
modified:   tests/NLog.UnitTests/Targets/Wrappers/RetryingTargetWrapperTests.cs
modified:   tests/NLog.UnitTests/TimeSourceTests.cs
modified:   tests/NLog.UnitTests/packages.config
modified:   tests/SampleExtensions/SampleExtensions.netfx35.csproj
modified:   tests/SampleExtensions/SampleExtensions.netfx45.csproj
modified:   tests/SampleExtensions/SampleExtensions.sl5.csproj
modified:   tests/SilverlightApp/SilverlightApp.sl5.csproj
modified:   tools/DumpApiXml/DocFileBuilder.cs
modified:   tools/DumpApiXml/DumpApiXml.csproj
modified:   tools/DumpApiXml/app.config
modified:   tools/MakeNLogXSD/MakeNLogXSD.csproj
modified:   tools/MakeNLogXSD/app.config
modified:   tools/MergeApiXml/MergeApiXml.csproj
modified:   tools/MergeApiXml/app.config

304NotModified commented:
I changed the line endings in the master. I expected less changes in this PR, but I was wrong I see

304NotModified commented:
Sorry that this PR wont merge now.

williamb1024 commented:
No problem, I'll see if I can figure out how rebase it on the EOL line normalization commit.

williamb1024 commented:
OK, I've merged the changes from master into my branch, so this should merge cleanly now.

304NotModified commented:
> NLog.UnitTests.LogFactoryTests.Auto_reload_validxml_test [FAIL] 

That's also a race condition :(

304NotModified commented:
Line endings looks good. I will review the code

304NotModified commented on src/NLog/GlobalDiagnosticsContext.cs:
```diff
@@ -57,22 +57,42 @@ public static void Set(string item, string value)
         /// Gets the Global Diagnostics Context named item.
         /// </summary>
         /// <param name="item">Item name.</param>
         /// <returns>The item value of string.Empty if the value is not present.</returns>
```
Can you fix the grammar error (it was already there): of => o**r**

williamb1024 replied:
Reworded it completely, now reads: <returns>The value of <paramref name="item"/>, if defined; otherwise <see cref="String.Empty"/>.</returns>

I believe there may be a few more of these floating around, I'll check and update them as well.

304NotModified commented on src/NLog/GlobalDiagnosticsContext.cs:
```diff
@@ -111,5 +131,17 @@ public static void Clear()
             }
         }
+
+        internal static string RenderContextObject ( object o )
```
I like the solution, but I was confused by the methode name. Proposal: `ConvertToString`

williamb1024 replied:
Sure. I chose `RenderContextObject` because I assumed at some point in the future it was likely that objects would have custom renderers. But the name change will work in either context.

304NotModified commented on src/NLog/GlobalDiagnosticsContext.cs:
```diff
@@ -57,22 +57,42 @@ public static void Set(string item, string value)
         /// <returns>The item value of string.Empty if the value is not present.</returns>
         public static string Get(string item)
         {
+            return RenderContextObject(GetObject(item));
```
I think this is not fully backwards compatible?

a item with the value `null` now returns `null` and previously `string.Empty`?

williamb1024 replied:
No, a `null` value will always return `String.Empty`. `String.Format` returns an empty string for "{0}" if the format argument is `null`.

I wasn't sure initially, so I tested it before writing the code this way.

304NotModified replied:
Yes you are right.

304NotModified commented on src/NLog/LayoutRenderers/GdcLayoutRenderer.cs:
```diff
@@ -57,8 +57,8 @@ public class GdcLayoutRenderer : LayoutRenderer
-            string msg = GlobalDiagnosticsContext.Get(this.Item);
-            builder.Append(msg);
+            // use AppendFormat to ensure that any IFormatProvider is respected
+            builder.AppendFormat(logEvent.FormatProvider, "{0}", GlobalDiagnosticsContext.GetObject(this.Item));
```
This looks a bit strange. (string.format and "{0}")

I propose: `GlobalDiagnosticsContext.Get(this.item, logEvent.FormatProvider). - (so a new overload)

williamb1024 replied:
As suggested.

304NotModified commented on src/NLog/LayoutRenderers/MdcLayoutRenderer.cs:
```diff
@@ -57,8 +58,8 @@ public class MdcLayoutRenderer : LayoutRenderer
-            string msg = MappedDiagnosticsContext.Get(this.Item);
-            builder.Append(msg);
+            // use AppendFormat to ensure that any IFormatProvider is respected
+            builder.AppendFormat(logEvent.FormatProvider, "{0}", MappedDiagnosticsContext.GetObject(this.Item));
```
idem as GlobalDiagnosticsContext. 

Too bad of the code duplication :|

williamb1024 replied:
As suggested.

304NotModified commented on src/NLog/LayoutRenderers/NdcLayoutRenderer.cs:
```diff
@@ -77,7 +77,7 @@ public NdcLayoutRenderer()
         protected override void Append(StringBuilder builder, LogEventInfo logEvent)
         {
-            string[] messages = NestedDiagnosticsContext.GetAllMessages();
+            object[] messages = NestedDiagnosticsContext.GetAllObjects();
```
GetAllMessages() should be better over here? (returning `string[]`)

williamb1024 replied:
Added a GetAllMessages() overload that accepts an `IFormatProvider` and used it.

304NotModified commented on src/NLog/MappedDiagnosticsContext.cs:
```diff
@@ -63,20 +63,39 @@ public static void Set(string item, string value)
         {
-            string s;
+            object o = GetObject(item);
+            return (o != null) ? o.ToString() : String.Empty;
```
Same possible backwardscomp problem

williamb1024 replied:
I missed this when adding RenderContextObject. It has been corrected to use ConvertToString now.

304NotModified commented:
Looks nice overall.  :+1:  

I added some small notes.

304NotModified commented:
I think we should expand the current documentation about the GDC, MDC and NDC

And I think over here:

- https://github.com/NLog/NLog/wiki/Layout-Renderers#passing-custom-values-to-a-layout
- https://github.com/NLog/NLog/wiki/Ndc-Layout-Renderer
- https://github.com/NLog/NLog/wiki/mdc-Layout-Renderer
- https://github.com/NLog/NLog/wiki/gdc-Layout-Renderer

I would be great if we advise when users should use NDC, MDC, GDC or event-properties.

Can you help on the documentation?

304NotModified commented:
@williamb1024 do you have time to elaborate the documentation? 

We are trying to merge only if the unit tests and docs are ready. (unit tests would be nice, but not a requirement at this PR)

williamb1024 commented:
Yes, I can assist with the documentation. I'll see what I can do over the next few days.

williamb1024 commented:
I've looked at the pages you listed previously. Any chance you can give me some direction on what you'd like to see. Honestly, documentation is not my strong suit, so a few pointers would be helpful.

304NotModified commented:
yeah sure,

I propose:

- https://github.com/NLog/NLog/wiki/Layout-Renderers#passing-custom-values-to-a-layout
 - I think this should mention the NDC/MDC/GDC. I should explain (in short) why/when we should use event-properties/NDC/MDC/GDC
- https://github.com/NLog/NLog/wiki/Ndc-Layout-Renderer
 - Needs explanation why/when we should use it.
 - Needs some examples.
- https://github.com/NLog/NLog/wiki/mdc-Layout-Renderer
 - Needs explanation why/when we should use it.
 - Needs some examples.
- https://github.com/NLog/NLog/wiki/gdc-Layout-Renderer
 - Needs explanation why/when we should use it.
 - Needs some examples.

304NotModified commented:
I will merge this if we expanded the docs

williamb1024 commented:
Been busy with work, I'll start updating the documentation this weekend.

304NotModified commented:
Thanks in advance!

williamb1024 commented:
I've modified the MDC layout renderer wiki page. Please review it and let me know if that is what you are looking for documentation wise. I'll continue with the other pages once you've validated.

304NotModified commented:
Looks very nice. 2 minor points:

- 4.0.1 should be 4.1 (or 4.1.0)
- the example is maybe a bit to complicated. The goal is to learn the usage of the mdc.  Other info (new classes etc)  can be a bit noisy.  Maybe be it's better to use standard. Net concepts/classes. What do you think?

williamb1024 commented:
I've added a set of more simplistic samples to the MDC and extended the GDC documentation page. I believe that the ability to create "on-demand" or "dynamic properties" is the key benefit to accepting `object` type parameters for the GDC, MDC and NDC, so it should be demonstrated in the examples.

I can't think of any out of the box .NET classes that would demonstrate that. I could create a sample that provides a `Func<string>` wrapper that calls the delegate as the `ToString()` implementation and then use that as a way to demonstrate the same functionality, but that would result in even more code.

304NotModified commented:
I'm thinking of another example for the dynamic properties. It's a funny one, because IMO it's a work around that we can set `Func` values. Do you agree?

We can although fix that with an helper class (almost the same to the `MdcTickProperty` class and some c# extensions. That saves use a difficult explanation ;)

304NotModified commented:
I added to the MDC a section which links to GDC and event properties. 

I cannot explain when to use the NDC.... do you know a good example?

williamb1024 commented:
I think you are suggesting the same thing as my `Func<string>` wrapper. Something like:

```
public class PropertyWrapper 
{
   private Func<string> _getter;

   public PropertyWrapper ( Func<string> getter )
   {
     // should validate parameter..
      _getter = getter;
   }

  public override string ToString () 
  {
      return _getter();
  }
}

MappedDiagnosticsContext.Set("Property", new PropertyWrapper(() => Environment.TickCount.ToString());
```

304NotModified commented:
yes! The `PropertyWrapper` is cool!

304NotModified commented:
I think the `func` should be `func<object>` ;)

williamb1024 commented:
Why?

williamb1024 commented:
I haven't thought of a good NDC example yet. Still thinking though.

304NotModified commented:
I think the why is the `func<object>` response.

It was for including `PropertyWrapper` in NLog. And because we all doing now proposing the `tostring`, that should be the case for the `func` also.

williamb1024 commented:
If you want to add it to the code directly, it would make more sense (to me anyway) to add "Func<object>" overloads to `Set` methods for the MDC and GDC and one for (I think) `Push` on the NDC.

304NotModified commented:
I have to leave, it's here after 1:00 AM... And work tomorrow (fulltime). I will respond tomorrow

304NotModified commented:
I think you right.

 There is one (indirect) benefit of the extension method, we need a shared interface (for mdc and gdc),  which is a good thing.  

I would be great if we could add the func methods to NLog :)

304NotModified commented:
Hi @williamb1024 

Would you to include the func overloads to the PR?

williamb1024 commented:
Yes but it will be a few days before I can get to it. Have been away for work.


On Aug 4, 2015, at 3:59 PM, Julian Verdurmen <notifications@github.com<mailto:notifications@github.com>> wrote:


Hi @williamb1024<https://github.com/williamb1024>

Would you to include the func overloads to the PR?

—
Reply to this email directly or view it on GitHub<https://github.com/NLog/NLog/pull/743#issuecomment-127759780>.

304NotModified commented:
Hi. Should we move this feature to 4.2?  

I really like to release 4.1.

williamb1024 commented:
That's fine, I will get to this as soon as possible, I've just had a number of other things going on that are outside of my control.

304NotModified commented:
- [ ] update docs to change 4.1 to 4.2

304NotModified commented:
note: I found a nice example for NDC:

http://stackoverflow.com/questions/334367/when-to-use-nested-diagnostic-context-ndc

We need that on the docs! If even didn't knew when to use the NDC

304NotModified commented:
Let merge this one, for the other things I have created new issues.

304NotModified commented:
docs are OK.
